lib/defaultMetrics.js

```diff
@@ -31,57 +33,54 @@ const metrics = {
 };
 const metricsList = Object.keys(metrics);
 
-let existingInterval = null;
-// This is used to ensure the program throws on duplicate metrics during first run
-// We might want to consider not supporting running the default metrics function more than once
-let init = true;
-
 module.exports = function startDefaultMetrics(config) {
-	let normalizedConfig = config;
-	if (typeof config === 'number') {
-		printDeprecationCollectDefaultMetricsNumber(config);
-
-		normalizedConfig = { timeout: config };
+	if (config !== null && config !== undefined && !isObject(config)) {
+		throw new Error('config must be null, undefined, or an object');
 	}
 
-	normalizedConfig = Object.assign(
+	config = Object.assign(
 		{
 			timestamps: true,
 			timeout: 10000
 		},
-		normalizedConfig
+		config
 	);
 
-	if (existingInterval !== null) {
-		clearInterval(existingInterval);
-	}
+	const registry = config.register || globalRegistry;
+	const last = registry
+		.collectors()
+		.find(collector => collector._source === metrics);
 
-	const initialisedMetrics = metricsList.map(metric => {
-		const defaultMetric = metrics[metric];
-		if (!init) {
-			defaultMetric.metricNames.map(
-				globalRegistry.removeSingleMetric,
-				globalRegistry
-			);
-		}
+	if (last) {
+		throw new Error(
+			'Cannot add the default metrics twice to the same registry'
+		);
+	}
 
-		return defaultMetric(normalizedConfig.register, normalizedConfig);
+	const scrapers = metricsList.map(key => {
+		const metric = metrics[key];
+		return metric(config.register, config);
 	});
 
-	function updateAllMetrics() {
-		initialisedMetrics.forEach(metric => metric.call());
+	// Ideally the library would be based around a concept of collectors and
+	// async callbacks, but in the short-term, trigger scraping of the
+	// current metric value synchronously.
+	// - // https://prometheus.io/docs/instrumenting/writing_clientlibs/#overall-structure
+	function defaultMetricCollector() {
+		scrapers.forEach(scraper => scraper());
 	}
 
-	updateAllMetrics();
-
-	existingInterval = setInterval(
-		updateAllMetrics,
-		normalizedConfig.timeout
-	).unref();
+	// defaultMetricCollector has to be dynamic, because the scrapers are in
+	// its closure, but we still want to identify a default collector, so
+	// tag it with a value known only to this module (the const metric array
+	// value) so we can find it later.
+	defaultMetricCollector._source = metrics;
+	registry.registerCollector(defaultMetricCollector);
 
-	init = false;
+	// Because the tests expect an immediate collection.
+	defaultMetricCollector();
 
-	return existingInterval;
+	return 0; // Backwards compat, used to be a timeout in milliseconds.
```

Comment:
no need for this I think, it's landing in a major anyways. Just returning nothing seems fine